File: go/internal/base64url/base64url.go
package base64url

import (
	"encoding/base64"
	"errors"
	"strings"
)

// DecodeStrict decodes a base64url encoded string without padding.
// It strictly rejects strings containing '=' padding characters or invalid URL-safe characters.
func DecodeStrict(s string) ([]byte, error) {
	if strings.Contains(s, "=") {
		return nil, errors.New("base64url string contains '=' padding, which is not allowed")
	}

	decoded, err := base64.RawURLEncoding.DecodeString(s)
	if err != nil {
		return nil, err
	}

	return decoded, nil
}

File: rust/src/base64url.rs
use base64::{Engine, engine::general_purpose::URL_SAFE_NO_PAD};

/// Decodes a base64url encoded string strictly without padding.
/// It rejects strings containing '=' padding characters or invalid URL-safe characters.
pub fn decode_strict(s: &str) -> Result<Vec<u8>, String> {
    if s.contains('=') {
        return Err("base64url string contains '=' padding, which is not allowed".to_string());
    }

    URL_SAFE_NO_PAD
        .decode(s)
        .map_err(|e| format!("Base64url decode error: {}", e))
}
